notebook/gateway/handlers.py

```diff
@@ -153,10 +153,16 @@ def _connect(self, kernel_id):
         self.ws_future.add_done_callback(self._connection_done)
 
     def _connection_done(self, fut):
-        if not self.disconnected and fut.exception() is None:  # prevent concurrent.futures._base.CancelledError
+        exc = fut.exception()
+        if not self.disconnected and exc is None:  # prevent concurrent.futures._base.CancelledError
             self.ws = fut.result()
             self.log.debug("Connection is ready: ws: {}".format(self.ws))
         else:
+            if exc is not None and isinstance(exc, HTTPClientError):
+                self.log.warning("Encountered this HTTPError: {}. Disconnecting if it's a 404.".format(exc))
+                if exc.code == 404:
+                    # Disconnect if the kernel no longer exists
+                    self._disconnect()
```

Comment:
commenting here since github won't let me comment outside edited lines. Basically, this kind of logic could maybe go on line 200, or I could add exponential backoff there.

Reply:
> or I could add exponential backoff there

Could you please add this?